### PR TITLE
Network Tests: bind host network udp listeners to hostIPs

### PR DESCRIPTION
**Pull request by fedepaol**
**What type of PR is this?**

/kind failing-test

**What this PR does / why we need it**:

When listening on udp, the reply is sent using a src address which is the address of the gateway interface. This means that when listening to any, the reply can be sent out with a src ip which is different from the request's target ip. This confuses natting and "connection-oriented" udp services do not work (see https://github.com/kubernetes/kubernetes/issues/95565 ).
Here, we force the endpoint to listen from the hostIP and from podIPs, to cover both dual stack and legacy clusters.

**Note:** needs the changes implemented in https://github.com/kubernetes/kubernetes/pull/98883 and the agnhost image to be bumped up.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by fedepaol** [comment]
/hold
This needs https://github.com/kubernetes/kubernetes/pull/98883 to be bumped

File: test/e2e/framework/network/utils.go [hunk 1]
```diff
@@ -540,6 +540,18 @@ func (config *NetworkingTestConfig) executeCurlCmd(cmd string, expected string)
...
+	netexecArgs := []string{
+		"netexec",
+		fmt.Sprintf("--http-port=%d", EndpointHTTPPort),
+		fmt.Sprintf("--udp-port=%d", EndpointUDPPort),
+	}
+	// In case of hostnetwork endpoints, we want to bind the udp listener to specific ip addresses.
+	// In order to cover legacy AND dualstack, we pass both the host ip and the two pod ips. Agnhost
+	// removes duplicates and so this will listen on both addresses (or on the single existing one).
+	if config.EndpointsHostNetwork {
+		netexecArgs = append(netexecArgs, "--udp-bind-on=$(POD_IP0),$(POD_IP1),$(HOST_IP)")
```
Comment (aojea) [inline test/e2e/framework/network/utils.go:552]: I think we changed this flag

File: test/e2e/framework/network/utils.go [hunk 2]
```diff
@@ -602,6 +610,35 @@ func (config *NetworkingTestConfig) createNetShellPodSpec(podName, hostname stri
...
+					FieldRef: &v1.ObjectFieldSelector{
+						FieldPath: "status.podIPs[0].ip",
+					},
+				},
+			},
+			{
+				Name: "POD_IP1",
+				ValueFrom: &v1.EnvVarSource{
+					FieldRef: &v1.ObjectFieldSelector{
+						FieldPath: "status.podIPs[1].ip",
```
Comment (aojea) [inline test/e2e/framework/network/utils.go:636]: just curious, what happens if podIPs only has 1 element? is POD_IP1 empty?  can the downward api deal with it?

**Comment by aojea** [comment]
/ok-to-test
/hold
it depends on the agnhost image modifications

File: test/e2e/framework/network/utils.go (hunk 2 above)
Comment (fedepaol) [reply test/e2e/framework/network/utils.go:636]: I ingenuosly thought the env variable would be set as empty, but I will try setting it to 2 to see what happens

Comment (fedepaol) [reply test/e2e/framework/network/utils.go:636]: Actually, not sure what I tested yesterday, but choosing an item of the array won't work.
The pod is failing with `Invalid value: "status.podIPs[0]": error converting fieldPath: field label not supported: status.podIPs[0]`

What works, is using `status.podIPs` and fills the variable with a comma separated list of addresses.

**Review by spiffxp — APPROVED** [review]
/approve
/lgtm
/milestone v1.21
Ping if you need a re-lgtm

**Comment by spiffxp** [comment]
https://github.com/kubernetes/k8s.io/pull/1769 merged, looking for postsubmit to succeed

**Comment by spiffxp** [comment]
/retest
It succeeded https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/post-k8sio-image-promo/1369337442347257856

**Comment by aojea** [comment]
ping @aramase 
I think that this was the problem we had in the dual-stack-azure jobs,
it LGTM too, cancel the hold if CI passes 
:clap:

**Comment by aojea** [comment]
/retest

k8s.io/kubernetes/test/integration/apiserver/admissionwebhook: TestWebhookTimeoutWithWatchCache expand_less
--

**Comment by fedepaol** [comment]
@spiffxp / @aojea all green, mind re-lgtm-ing / un-holding?
Thanks!

**Comment by aojea** [comment]
This slipped the code-freeze windows, but is a test fix
@spiffxp knows better the procedures, I think we need a milestone set now

**Comment by spiffxp** [comment]
/milestone v1.21
Ok to land before test freeze

**Pull request merged**